blockchain/v2/scheduler.go

```diff
@@ -0,0 +1,715 @@
+			delete(sc.pendingBlocks, height)
+		}
+	}
+
+	for height, rcvPeerID := range sc.receivedBlocks {
+		if rcvPeerID == peerID {
+			sc.setStateAtHeight(height, blockStateNew)
+			delete(sc.receivedBlocks, height)
+		}
+	}
+
+	// remove the blocks from blockStates if the peer removal causes the max peer height to be lower.
+	peer.state = peerStateRemoved
+	maxPeerHeight := int64(0)
+	for _, otherPeer := range sc.peers {
+		if otherPeer.state != peerStateReady {
+			continue
+		}
+		if otherPeer.peerID != peer.peerID && otherPeer.height > maxPeerHeight {
+			maxPeerHeight = otherPeer.height
+		}
+	}
+	for h := range sc.blockStates {
+		if maxPeerHeight < h {
+			delete(sc.blockStates, h)
+		}
+	}
+
+	return nil
+}
+
+// check if the blockPool is running low and add new blocks in New state to be requested.
+// This function is called when there is an increase in the maximum peer height or when
+// blocks are processed.
+func (sc *scheduler) addNewBlocks() {
+	if len(sc.blockStates) >= sc.targetPending {
+		return
+	}
+
+	for i := sc.height; i < int64(sc.targetPending)+sc.height; i++ {
+		if i > sc.maxHeight() {
+			break
+		}
+		if sc.getStateAtHeight(i) == blockStateUnknown {
+			sc.setStateAtHeight(i, blockStateNew)
+		}
+	}
+}
+
+func (sc *scheduler) setPeerHeight(peerID p2p.ID, height int64) error {
+	peer, ok := sc.peers[peerID]
+	if !ok {
+		return fmt.Errorf("cannot find peer %s", peerID)
+	}
+
+	if peer.state == peerStateRemoved {
+		return fmt.Errorf("cannot set peer height for a peer in peerStateRemoved")
+	}
+
+	if height < peer.height {
+		return fmt.Errorf("cannot move peer height lower. from %d to %d", peer.height, height)
+	}
+
+	peer.height = height
+	peer.state = peerStateReady
+
+	sc.addNewBlocks()
+	return nil
+}
+
+func (sc *scheduler) getStateAtHeight(height int64) blockState {
+	if height <= sc.initHeight {
+		return blockStateProcessed
+	} else if state, ok := sc.blockStates[height]; ok {
+		return state
+	} else {
+		return blockStateUnknown
+	}
+}
+
+func (sc *scheduler) getPeersAtHeight(height int64) []p2p.ID {
+	peers := make([]p2p.ID, 0)
+	for _, peer := range sc.peers {
+		if peer.state != peerStateReady {
+			continue
+		}
+		if peer.height >= height {
+			peers = append(peers, peer.peerID)
+		}
+	}
+	return peers
+}
+
+func (sc *scheduler) peersInactiveSince(duration time.Duration, now time.Time) []p2p.ID {
+	peers := make([]p2p.ID, 0)
+	for _, peer := range sc.peers {
+		if peer.state != peerStateReady {
+			continue
+		}
+		if now.Sub(peer.lastTouched) > duration {
+			peers = append(peers, peer.peerID)
+		}
+	}
+	return peers
+}
+
+func (sc *scheduler) peersSlowerThan(minSpeed int64) map[p2p.ID]struct{} {
+	peers := make(map[p2p.ID]struct{})
+	for peerID, peer := range sc.peers {
+		if peer.state != peerStateReady {
+			continue
+		}
+		if peer.lastRate < minSpeed {
+			peers[peerID] = struct{}{}
+		}
+	}
+	return peers
+}
+
+// XXX: this duplicates the logic of peersInactiveSince and peersSlowerThan
```

Comment:
why not used them here then?

Reply:
I thought about replacing the specific filters with a function that took a predicate function but in the end it seems like it might not be a good bang for buck unless there are so many filter functions that we get tired of `for` loops. I've tweaked the code to make them more standardized.